expand_test.go

```diff
@@ -0,0 +1,244 @@
+		}
+
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	// Parse whole string
+	tr := transform.NewReader(r, &expandTransformer{expand: expand_func})
+	actual, err := ioutil.ReadAll(tr)
+	require.NoError(t, err)
+	assert.Exactly(t, expected, string(actual))
+
+	_, err = r.Seek(0, io.SeekStart)
+	require.NoError(t, err)
+
+	// Partial parse
+	var partial [11]byte
+	tr = transform.NewReader(r, &expandTransformer{expand: expand_func})
+	n, err := tr.Read(partial[:])
+	require.NoError(t, err)
+	assert.Exactly(t, n, len(partial))
+	assert.Exactly(t, expected[:n], string(partial[:]))
+
+	// Empty string
+	r = bytes.NewReader([]byte{})
+	tr = transform.NewReader(r, &expandTransformer{expand: expand_func})
+	actual, err = ioutil.ReadAll(tr)
+	require.NoError(t, err)
+	assert.Exactly(t, "", string(actual))
+}
+
+func TestExpanderOneByteAtATime(t *testing.T) {
+	var buf []byte
+
+	buf = append(buf, orig...)
+
+	r := bytes.NewReader(buf)
+	rr := &oneByteReader{r: r}
+
+	expand_func := func(s string) (string, error) {
+		switch s {
+		case "t3sT":
+			return "test", nil
+		case "i5_":
+			return "is ", nil
+		}
+
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	tr := transform.NewReader(rr, &expandTransformer{expand: expand_func})
+	actual, err := ioutil.ReadAll(tr)
+	require.NoError(t, err)
+	assert.Exactly(t, expected, string(actual))
+}
+
+func TestExpanderFailingTransform(t *testing.T) {
+	var buf []byte
+
+	buf = append(buf, orig...)
+
+	r := bytes.NewReader(buf)
+
+	expand_func := func(s string) (string, error) {
+		switch s {
+		case "t3sT":
+			return "test", nil
+		// missing "i5_" case
+		}
+
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	tr := transform.NewReader(r, &expandTransformer{expand: expand_func})
+	_, err := ioutil.ReadAll(tr)
+	require.Error(t, err)
+}
+
+func TestExpanderMisc(t *testing.T) {
+	tests := [...]struct {
+		orig   string
+		expect string
+	}{
+		{ends_in_dollar, ends_in_dollar},
+		{ends_in_ddollar, ends_in_dollar},
+		{ends_in_var, ends_in_var_expect},
+		{many_dollars_orig, many_dollars_expect},
+	}
+
+	expand_func := func(s string) (string, error) {
+		switch s {
+		case "t3sT":
+			return "test", nil
+			// missing "i5_" case
+		}
+
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	for _, tst := range tests {
+		var buf []byte
+		buf = append(buf, tst.orig...)
+		tr := transform.NewReader(
+			bytes.NewReader(buf),
+			&expandTransformer{expand: expand_func},
+		)
+		actual, err := ioutil.ReadAll(tr)
+		require.NoError(t, err)
+		assert.Exactly(t, tst.expect, string(actual))
+	}
+}
+
+func TestLongSrc(t *testing.T) {
+	a := strings.Repeat("a", transformBufSize-1)
+
+	tests := [...]struct {
+		orig   string
+		expect string
+	}{
+		{"foo${a}" + a, "foo" + a + a},
+		{"${a}foo$a", a + "foo" + a},
+		{"$a${", a + "${"},
+	}
+
+	expand_func := func(s string) (string, error) {
+		switch s {
+		case "a":
+			return a, nil
+		}
+
+		return "NOMATCH", errors.New("No Match")
+	}
+
+	for _, tst := range tests {
```

Comment:
Could you use [subtests](https://blog.golang.org/subtests) instead of table driven tests?

Reply:
Sure.